internal/runtime/storeforward.go

```diff
@@ -0,0 +1,188 @@
+//
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"runtime"
+	"time"
+
+	"github.com/edgexfoundry/app-functions-sdk-go/appcontext"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/common"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/contracts"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db/interfaces"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+)
+
+const (
+	DefaultMinRetryInterval = 5000 // 5 secs
+)
+
+type storeForwardInfo struct {
+	runtime      *GolangRuntime
+	storeClient  interfaces.StoreClient
+	pipelineHash string
+}
+
+func (sf *storeForwardInfo) startStoreAndForwardRetryLoop(appCtx context.Context, enabledCtx context.Context, serviceKey string, config *common.ConfigurationStruct, edgeXClients common.EdgeXClients) {
+
+	go func() {
+
+		retryInterval := config.Writable.StoreAndForward.RetryInterval
+		if retryInterval < DefaultMinRetryInterval {
+			retryInterval = DefaultMinRetryInterval
+		}
+
+		if config.Writable.StoreAndForward.MaxRetryCount < 0 {
+			edgeXClients.LoggingClient.Info(fmt.Sprintf("StoreAndForward MaxRetryCount can not be less than 0, defaulting to 1"))
+			config.Writable.StoreAndForward.MaxRetryCount = 1
+		}
+
+		edgeXClients.LoggingClient.Info(fmt.Sprintf("Starting StoreAndForward Retry Loop with %d RetryInterval and %d max retries", retryInterval, config.Writable.StoreAndForward.MaxRetryCount))
+
+	exit:
+		for {
+			select {
+
+			case <-appCtx.Done():
```

Comment:
What if appCtx is nil?

Reply:
This isn't in a public API, so we control that it is never nil. Similar reason to why your BootStraphandlers aren't checking the ctx for nil. :-)

Reply:
I called it out because I wasn't able to confirm that it's never nil based on the code in this PR.